### PR TITLE
Moved and refactored open-editor-state; closes #1953.

### LINKED ISSUE #1953: Extract Open Editor State Redux Logic
- copy logic from `/assets/ZZZ/data/eventespresso/open-editor-state/index.js` and move it into `/assets/prototype/application/ui/components/layout/editor-modal` 
- convert logic to implement `useReducer` instead of Redux
- rename file/function to something like `useOpenEditorState`
- find all prototype components that control the display of `assets/prototype/application/ui/components/forms/FormModal.js` (such as `assets/prototype/domain/eventEditor/ticketsList/ticketCard/EditTicketButton.js`) and convert them to use `useOpenEditorState` instead of their own local state

### PR DESCRIPTION
**Pull request by nicolad**
## Problem this Pull Request solves
Implements: https://github.com/eventespresso/event-espresso-core/issues/1953

## Checklist

* [ ] I have read the documentation relating to systems affected by this pull request, see https://github.com/eventespresso/event-espresso-core/tree/master/docs
* [ ] User input is adequately validated and sanitized
* [ ] all publicly displayed strings are internationalized (usually using `esc_html__()`, see https://codex.wordpress.org/I18n_for_WordPress_Developers)
* [ ] My code is tested.
* [ ] My code follows the Event Espresso code style.
* [ ] My code has proper inline documentation.
* [ ] My code accounts for when the site is in Maintenance Mode (MM2 especially disallows usage of models)

### CODE REVIEW

File: assets/prototype/domain/eventEditor/tickets/ticketPriceCalculator/TicketPriceCalculatorButton.js [hunk 1]
```diff
@@ -1,24 +1,21 @@
...
-	const handleClose = (e) => {
-		e.preventDefault();
-		e.stopPropagation();
-		setIsOpen(false);
-	};
+	const { getIsOpen, onClose, setIsOpen } = useContext(TicketContext);
+	const modalId = 'TicketPriceCalculator';
+	const isOpen = getIsOpen(modalId);
+	const onOpen = () => setIsOpen(modalId);
+	const handleClose = () => onClose(modalId);
```
Comment (nicolad) [inline assets/prototype/domain/eventEditor/tickets/ticketPriceCalculator/TicketPriceCalculatorButton.js]: This is needed to discriminate between calculator modal and edit modal.

**Comment by nicolad** [comment]
Created a followup issue for adding more related types: https://github.com/eventespresso/event-espresso-core/projects/6#card-30944214

File: assets/prototype/application/ui/components/layout/editor-modal/useEditorModalState.ts [hunk 2]
```diff
@@ -0,0 +1,97 @@
...
+	editorId?: string;
+	type: ActionType;
+};
+
+const reducer = (state: State, action: Action) => {
+	const { editorId } = action;
+
+	switch (action.type) {
+		case CLOSE_EDITOR:
+			if (editorId === head(state)) {
```
Comment (manzoorwanijk) [inline assets/prototype/application/ui/components/layout/editor-modal/useEditorModalState.ts]: How is it guaranteed that `editorId` will always be the first element of the object?

File: assets/prototype/application/ui/components/layout/editor-modal/useEditorModalState.ts [hunk 3]
```diff
@@ -0,0 +1,97 @@
...
+	}, [state]);
+
+	const openEditor = (editorId) => {
+		dispatch({
+			type: OPEN_EDITOR,
+			editorId,
+		});
+	};
+
+	const closeEditor = (editorId) => {
```
Comment (manzoorwanijk) [inline assets/prototype/application/ui/components/layout/editor-modal/useEditorModalState.ts]: Better to make it `editorId: string`

File: assets/prototype/domain/eventEditor/tickets/ticketPriceCalculator/TicketPriceCalculatorButton.js (hunk 1 above)
Comment (manzoorwanijk) [inline assets/prototype/domain/eventEditor/tickets/ticketPriceCalculator/TicketPriceCalculatorButton.js]: Nice

File: assets/prototype/application/ui/components/layout/editor-modal/useEditorModalState.ts (hunk 2 above)
Comment (nicolad) [reply assets/prototype/application/ui/components/layout/editor-modal/useEditorModalState.ts]: Not sure, this was the previous implementation: https://github.com/eventespresso/event-espresso-core/blob/EDTR/master/assets/ZZZ/data/eventespresso/open-editor-state/index.js#L31

Comment (nicolad) [reply assets/prototype/application/ui/components/layout/editor-modal/useEditorModalState.ts]: Maybe there was a reason for that.
@tn3rb what was the intention there?

File: assets/prototype/application/ui/components/layout/editor-modal/useEditorModalState.ts (hunk 3 above)
Comment (nicolad) [reply assets/prototype/application/ui/components/layout/editor-modal/useEditorModalState.ts]: Added: https://github.com/eventespresso/event-espresso-core/pull/2074/commits/481f0310ad52b9bad886b498d4073c545af54e8d
Thx :bowing_man:

File: assets/prototype/application/ui/components/layout/editor-modal/useEditorModalState.ts (hunk 2 above)
Comment (manzoorwanijk) [reply assets/prototype/application/ui/components/layout/editor-modal/useEditorModalState.ts]: OK, may be knows the reason

**Pull request merged**